_posts/2020-04-13-Julia-Antipatterns.md

```diff
@@ -0,0 +1,831 @@
+For more complicated objects it can be quiet large, often larger than `isequal` as designing a good `hash` function is hard.
+But the case we see in this antipattern are mostly `Dict{Symbol}` or `Dict{String}`
+
+**Input:**
+
+<div class="jupyter-input jupyter-cell">
+
+```julia
+dict = Dict([:a=>1, :b=>2, :c=>3, :d=>4, :e=>5])
+@btime $(dict)[:d];
+```
+</div>
+
+**Output:**
+
+<div class="jupyter-stream jupyter-cell">
+
+```
+
+  8.456 ns (0 allocations: 0 bytes)
+
+```
+</div>
+
+**Input:**
+
+<div class="jupyter-input jupyter-cell">
+
+```julia
+str_dict = Dict(string(k)=> v for (k,v) in dict)  # convert all the keys to strings
+@btime $(str_dict)["d"];
+```
+</div>
+
+**Output:**
+
+<div class="jupyter-stream jupyter-cell">
+
+```
+
+  18.077 ns (0 allocations: 0 bytes)
+
+```
+</div>
+
+One alternative is [OrderedCollection's](https://github.com/JuliaCollections/OrderedCollections.jl/) `LittleDict`.
+This is a naive pair of lists based dictionary, with expected time `O(n)` but with a much lower cost per actual element as it doesn't need to `hash` anything.
+
+**Input:**
+
+<div class="jupyter-input jupyter-cell">
+
+```julia
+using OrderedCollections
+little_dict = LittleDict(dict)
+@btime $(little_dict)[:d];
+```
+</div>
+
+**Output:**
+
+<div class="jupyter-stream jupyter-cell">
+
+```
+
+  4.793 ns (0 allocations: 0 bytes)
+
+```
+</div>
+
+It also comes in a immutable form via `freeze` (or by constructing it with `Tuple`s)
+
+**Input:**
+
+<div class="jupyter-input jupyter-cell">
+
+```julia
+frozen_little_dict = freeze(LittleDict(dict))
+@btime $(frozen_little_dict)[:d];
+```
+</div>
+
+**Output:**
+
+<div class="jupyter-stream jupyter-cell">
+
+```
+
+  4.547 ns (0 allocations: 0 bytes)
+
+```
+</div>
+
+But the real winner here is the `NamedTuple`.
+Which performs [constant-folding](https://en.wikipedia.org/wiki/Constant_folding) to remove the indexing operation entirely, and just put the result directly into the AST during compilation.
+
+**Input:**
+
+<div class="jupyter-input jupyter-cell">
+
+```julia
+named_tuple = (; dict...)  # create a NamedTuple with the same content as the dict
+@btime $(named_tuple)[:d];
+```
+</div>
+
+**Output:**
+
+<div class="jupyter-stream jupyter-cell">
+
+```
+
+  0.032 ns (0 allocations: 0 bytes)
+
+```
+</div>
+
+If one is thinking _"I have some constants and I want to group them"_ then look no further than the `NamedTuple`.
+It is the right answer.
+`Dict` is best when you don't know all the keys when you write the code, and/or if values need to be updated.
+
+## Over constraining argument types
+
+Type constraints in Julia are **only** for dispatch.
+If you don't have multiple methods for a function, you don't need any type-constraints.
+If you must add type-constraints (for dispatch) do so as loosely as possible.
+
+### Reasons people do this:
+I think this comes from three main places.
+The belief it would make the code faster (false), safer (mostly false), or easier to understand (true)
```

Comment:
> or easier to understand (true)

here you indicate clearly that there is at least one reason to add some type constraints, which contradicts the initial claim of:
> Type constraints in Julia are **only** for dispatch.

Reply:
they are only for dispatch.
The fact that they can do other things does not change their purpose.

I can use a hammer as a paper weight, but that does not change the purpose of the hammer to be "beating in nails and holding down paper".